core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -668,15 +667,25 @@ public static TimeValue parseTimeValue(String s, TimeValue minValue, String key)
     }
 
     public static Setting<Boolean> boolSetting(String key, boolean defaultValue, Property... properties) {
-        return new Setting<>(key, (s) -> Boolean.toString(defaultValue), Booleans::parseBooleanExact, properties);
+        return new Setting<>(key, (s) -> Boolean.toString(defaultValue), (value) -> parseBoolean(key, value), properties);
     }
 
     public static Setting<Boolean> boolSetting(String key, Setting<Boolean> fallbackSetting, Property... properties) {
-        return new Setting<>(key, fallbackSetting, Booleans::parseBooleanExact, properties);
+        return new Setting<>(key, fallbackSetting, (value) -> parseBoolean(key, value), properties);
     }
 
     public static Setting<Boolean> boolSetting(String key, Function<Settings, String> defaultValueFn, Property... properties) {
-        return new Setting<>(key, defaultValueFn, Booleans::parseBooleanExact, properties);
+        return new Setting<>(key, defaultValueFn, (value) -> parseBoolean(key, value), properties);
+    }
+
+    private static Boolean parseBoolean(String key, String value) {
+        // let the parser handle all cases for non-proper booleans without a deprecation warning by throwing IAE
+        boolean booleanValue = Booleans.parseBooleanExact(value);
+        if (Booleans.isStrictlyBoolean(value) == false) {
+            DeprecationLogger deprecationLogger = new DeprecationLogger(Loggers.getLogger(Setting.class));
+            deprecationLogger.deprecated("Expected a boolean for setting [{}] but got [{}]", key, value);
```

Comment:
I see this is all over the place, so if you don't want to add it, that's okay with me. It's up to you :)